src/validation/validation.cpp

```diff
@@ -2833,28 +2833,15 @@ bool DisconnectTip(CValidationState &state, const Consensus::Params &consensusPa
     // rolling back the chain using the "rollbackchain" rpc command.
     if (!fRollBack)
     {
-        std::vector<uint256> vHashUpdate;
         for (const auto &ptx : block.vtx)
         {
-            // ignore validation errors in resurrected transactions
-            std::list<CTransactionRef> removed;
-            CValidationState stateDummy;
-            if (ptx->IsCoinBase() ||
-                !AcceptToMemoryPool(mempool, stateDummy, ptx, AreFreeTxnsDisallowed(), nullptr, true))
-            {
-                mempool.removeRecursive(*ptx, removed);
-            }
-            else if (mempool.exists(ptx->GetHash()))
-            {
-                vHashUpdate.push_back(ptx->GetHash());
-            }
+            CTxInputData txd;
+            txd.tx = ptx;
+            txd.nodeId = -1;
+            txd.nodeName = "rollback";
+            txd.whitelisted = false;
+            EnqueueTxForAdmission(txd);
```

Comment:
The main problem I see with this I that It may impact the wallet balance in some way...when disconnecting the tip the wallet has to be sync'd as you go backward but here the wallet sync won't happen until potential the rollback happened and possibly the new tip established...seems like some sort of wallet conflict could occur here due to timing.

also minor bug -> you don't have to enqueue the coinbase tx.

Reply:
Thx: fixed the coinbase tx.

I didn't remove the SyncWithWallets call that you'll see a few lines down.  But I think that you are right that the order is incorrect in a multithreaded context.  Its not effectively multithreaded right now (cs_main is taken), but we might as well fix this now.  Basically, the current order is to re-enqueue the tx (which in a multithread context means they could start being evaluated), then update the tip, then remove from the wallet.  But we need to update the tip and remove from the wallet BEFORE we re-enqueue these tx.  Otherwise they may be rejected from the mempool as double spent (by themselves since the unwind is not finished).